src/cheri_insts.sail

```diff
@@ -1206,12 +1206,7 @@ function handle_store_cap_via_cap(rs, cs, cap_val, vaddrBits) = {
   } else if not(is_aligned_addr(vaddrBits, cap_size)) then {
     handle_mem_exception(vaddrBits, E_SAMO_Addr_Align());
     RETIRE_FAIL
-  } else match translateAddr(vaddrBits, Write(Cap)) {
-    TR_Failure(E_Extension(EXC_CHERI_VMEM_STORE_CAP), _) => {
-      handle_cheri_cap_exception(CapEx_TLBNoStoreCap, cs);
-      RETIRE_FAIL
-    },
-    TR_Failure(E_Extension(_), _) => { internal_error("unexpected cheri exception for cap store") },
+  } else match translateAddr(vaddrBits, Write(if rs_val.tag then Cap else Data)) {
```

Comment:
I've left similar branching on `rs_val.tag` in #18, but the more I think about it the more I'm convinced it's not the correct approach.  It doesn't work for amoswap.c, as there's but one flag to convey both load and store.  This suggests that we should always be passing `Cap` in and that it's up to the PTW logic, not `cheri_insts`, to decide whether we're operating in a data-dependent way (by returning a caveated `TR_Address`) or -independent (by returning `TR_Failure`).

As we don't have amoswap.c yet and I don't wish to impede progress, I'm happy for this to land as it stands and we can shake out the details as part of #18.

Reply:
Yeah, that will need more work. This was just the minimal set of things needed to get it working today rather than in a week or month.

Reply:
Also, this approach _would_ work if we made ReadWrite take a pair; then you could have Cap always in the first and Data/Cap in the second for AMOSWAP.C. That then pushes the trapping logic into the PTW where it really belongs, leaving just tag clearing to the instruction. I think that's actually nicest?

Reply:
That's what I was proposing in https://github.com/rems-project/sail-riscv/pull/57

Reply:
I think I may have initially doubted that approach, but yes. Though we only want a subset of that; the r and w _types_ can be the same, just that you get two values in ReadWrite.